src/dioptra/restapi/errors.py

```diff
@@ -20,87 +20,391 @@
 """
 from __future__ import annotations
 
+import typing
+from io import StringIO
+
+import structlog
+from flask import request
 from flask_restx import Api
+from structlog.stdlib import BoundLogger
+
+LOGGER: BoundLogger = structlog.stdlib.get_logger()
+
+"""
+Helper function to add attribute name/value pairs to StringIO instance as an error
+message.
+Args:
+    buffer: The StringIO instance into which the attribute name/value pairs are to be
+    added.
+    attributes: the list of attribute value pairs to add to the buffer.
+"""
+
+
+def add_attribute_values(buffer: StringIO, **kwargs: typing.Any) -> None:
+    index = 0
+    length = len(kwargs)
+    for key, value in kwargs.items():
+        if index != 0:
+            buffer.write(", ")
+            if index == length - 1:
+                buffer.write("and ")
+        buffer.write(f"{key} having value ({value})")
+        index += 1
+
+
+class DioptraError(Exception):
+    """
+    Generic Dioptra Error.
+    Args:
+        message: An error specific message to display that provide context for why the
+        error was raised.
+    """
+
+    def __init__(self, message: str):
+        self.message: str = message
+
+    def to_message(self) -> str:
+        if self.__cause__ is not None:
+            if isinstance(self.__cause__, DioptraError):
+                self.message = f"{self.message} Cause: {self.__cause__.to_message()}"
+            else:
+                self.message = f"{self.message} Cause: {self.__cause__}"
+        return self.message
 
 
-class BackendDatabaseError(Exception):
+class SubmissionError(DioptraError):
+    """The submission input is in error.
+    Args:
+        type: the resource type
+        action: the action
+    """
+
+    def __init__(self, type: str, action: str):
+        super().__init__(f"Input Error while attempting to {action} for {type}.")
+        self.resource_type = type
+        self.action = action
+
+
+class EntityDoesNotExistError(DioptraError):
+    """
+    The requested entity does not exist.
+    Args:
+        entity_type: the entity type name (e.g. "group" or "queue")
+        kwargs: the attribute value pairs used to request the entity
+    """
+
+    def __init__(self, entity_type: str, **kwargs: typing.Any):
+        buffer = StringIO()
```

Comment:
Memory/Computational efficiency in the same way that StringBuilder in Java is more efficient than a series of concatenations. I also felt like it was easier to read.

Reply:
This style is a little unusual in Python world, but the point is taken that it might be more memory efficient. Computational efficiency I'm unsure of, CPython has optimizations that are biased in favor of list comprehensions, I'd be interested to do a couple of quick benchmarks to see if this proves out in practice.

If we're going to keep the StringIO, I'd recommend using it in a `with` context block:

```python
with StringIO() as buffer:
    buffer.write(f"Failed to locate {entity_type} with ")
    add_attribute_values(buffer, **kwargs)
    buffer.write(".")
    super().__init__(buffer.getvalue())

self.entity_type = entity_type
self.entity_attributes = kwargs
```

Reply:
added the with context block

Reply:
@hbooth @keithmanville I was curious and put together a benchmark script comparing the StringIO-based approach and a list comprehension approach. Here's the results from the benchmarking script I ran:

    Running benchmarks for BufferEntityDoesNotExistError and ListCompEntityDoesNotExistError
    10000 iterations per benchmark, 10 repeats, showing fastest time per num_kwargs
    num_kwargs = 1       buffer time = 0.01083129  listcomp time = 0.01064946
    num_kwargs = 10      buffer time = 0.03197296  listcomp time = 0.03117646
    num_kwargs = 100     buffer time = 0.24803596  listcomp time = 0.23951279
    num_kwargs = 1000    buffer time = 1.72688354  listcomp time = 1.70404604
    num_kwargs = 10000   buffer time = 2.78957425  listcomp time = 2.74814846

So, list comprehensions actually win out on computational efficiency it seems, although it's a very small difference (the times are the total time over 10000 iterations, not the time per iteration).

Here's the benchmarking script if you want to look at it:

```python
import timeit
import typing
from io import StringIO

import faker

FAKE = faker.Faker()
BENCHMARK_NUM_KWARGS = [1, 10, 100, 1000, 10000]


class DioptraError(Exception):
    """
    Generic Dioptra Error.
    Args:
        message: An error specific message to display that provide context for why the
        error was raised.
    """

    def __init__(self, message: str):
        self.message: str = message

    def to_message(self) -> str:
        if self.__cause__ is None:
            return self.message

        if isinstance(self.__cause__, DioptraError):
            return f"{self.message} Cause: {self.__cause__.to_message()}"

        return f"{self.message} Cause: {self.__cause__}"


class BufferEntityDoesNotExistError(DioptraError):
    """
    The requested entity does not exist.
    Args:
        entity_type: the entity type name (e.g. "group" or "queue")
        kwargs: the attribute value pairs used to request the entity
    """

    def __init__(self, entity_type: str, **kwargs: typing.Any):
        with StringIO() as buffer:
            buffer.write(f"Failed to locate {entity_type} with ")
            buffer_add_attribute_values(buffer, **kwargs)
            buffer.write(".")
            super().__init__(buffer.getvalue())
        self.entity_type = entity_type
        self.entity_attributes = kwargs


class ListCompEntityDoesNotExistError(DioptraError):
    """
    The requested entity does not exist.
    Args:
        entity_type: the entity type name (e.g. "group" or "queue")
        kwargs: the attribute value pairs used to request the entity
    """

    def __init__(self, entity_type: str, **kwargs: typing.Any):
        buffer = " ".join(
            [f"Failed to locate {entity_type} with", *add_attribute_values(**kwargs)]
        )
        super().__init__(buffer)
        self.entity_type = entity_type
        self.entity_attributes = kwargs


def buffer_add_attribute_values(buffer: StringIO, **kwargs: typing.Any) -> None:
    index = 0
    length = len(kwargs)
    for key, value in kwargs.items():
        if index != 0:
            buffer.write(", ")
            if index == length - 1:
                buffer.write("and ")
        buffer.write(f"{key} having value ({value})")
        index += 1


def add_attribute_values(**kwargs: typing.Any) -> list[str]:
    length = len(kwargs)

    def sep(index: int) -> str:
        if index == length - 1:
            return "."

        if index == length - 2:
            return ", and"

        return ","

    return [
        f"{key} having value ({value}){sep(index)}"
        for index, (key, value) in enumerate(kwargs.items())
    ]


def benchmark_setup(fake: faker.Faker, num_kwargs: int) -> dict[str, typing.Any]:
    fake = faker.Faker()
    kwargs = {}
    for _ in range(num_kwargs):
        kwargs[fake.word()] = fake.sentence()

    return kwargs


def run_buffer_benchmark(entity_type, **kwargs) -> str:
    e = BufferEntityDoesNotExistError(entity_type, **kwargs)
    return e.to_message()


def run_listcomp_benchmark(entity_type, **kwargs) -> str:
    e = ListCompEntityDoesNotExistError(entity_type, **kwargs)
    return e.to_message()


if __name__ == "__main__":
    num_repeats = 10
    num_iterations = 10000
    entity_type = FAKE.word()
    print(
        "Running benchmarks for BufferEntityDoesNotExistError and "
        "ListCompEntityDoesNotExistError"
    )
    print(
        f"{num_iterations} iterations per benchmark, {num_repeats} repeats, showing "
        "fastest time per num_kwargs"
    )
    for num_kwargs in BENCHMARK_NUM_KWARGS:
        kwargs = benchmark_setup(FAKE, num_kwargs)
        t_buffer = timeit.Timer(
            stmt="run_buffer_benchmark(entity_type, **kwargs)",
            setup="from __main__ import run_buffer_benchmark, entity_type, kwargs",
        )
        t_listcomp = timeit.Timer(
            stmt="run_listcomp_benchmark(entity_type, **kwargs)",
            setup="from __main__ import run_listcomp_benchmark, entity_type, kwargs",
        )
        result_buffer = t_buffer.repeat(num_repeats, number=num_iterations)
        result_t_listcomp = t_listcomp.repeat(num_repeats, number=num_iterations)
        print(
            f"num_kwargs = {str(num_kwargs):6s}  "
            f"buffer time = {round(min(result_buffer), 8):.8f}  "
            f"listcomp time = {round(min(result_t_listcomp), 8):.8f}"
        )
```